apps/champions/lib/champions/units.ex

```diff
@@ -207,6 +217,151 @@ defmodule Champions.Units do
       %CurrencyCost{currency_id: Currencies.get_currency_by_name!("Gems").id, amount: 50}
     ]
 
+  ##########
+  # Fusion #
+  ##########
+
+  @doc """
+  Consume a list of units that meet specific rank and character requirements based on the target
+  unit's rank in order to increase it.
+
+  Returns `{:ok, unit}` or `{:error, reason}`.
+  """
+  def fuse(user_id, unit_id, consumed_units_ids) do
+    with {:unit, {:ok, unit}} <- {:unit, Units.get_unit(unit_id)},
+         {:unit_owned, true} <- {:unit_owned, unit.user_id == user_id},
+         {:can_rank_up, true} <- {:can_rank_up, can_rank_up(unit)},
+         consumed_units <- Units.get_units_by_ids(consumed_units_ids),
+         {:consumed_units_count, true} <-
+           {:consumed_units_count, Enum.count(consumed_units) == Enum.count(consumed_units_ids)},
+         {:consumed_units_valid, true} <-
+           {:consumed_units_valid, validate_consumed_units(unit, consumed_units)} do
+      result =
+        Multi.new()
+        |> Multi.run(:unit, fn _, _ -> Units.add_rank(unit) end)
+        |> Multi.run(:deleted_units, fn _, _ -> delete_consumed_units(consumed_units_ids) end)
+        |> Transaction.run()
+
+      case result do
+        {:error, reason} ->
+          {:error, reason}
+
+        {:error, _, _, _} ->
+          {:error, :transaction}
+
+        {:ok, %{unit: unit}} ->
+          {:ok, unit}
+      end
+    else
+      {:unit, {:error, :not_found}} ->
+        {:error, :not_found}
+
+      {:unit_owned, false} ->
+        {:error, :not_owned}
+
+      {:can_rank_up, false} ->
+        {:error, :cant_rank_up}
+
+      {:consumed_units_count, false} ->
+        {:error, :consumed_units_not_found}
+
+      {:consumed_units_valid, false} ->
+        {:error, :consumed_units_invalid}
+    end
+  end
+
+  defp delete_consumed_units(unit_ids) do
+    {amount_deleted, _return} = Units.delete_units(unit_ids)
+
+    if Enum.count(unit_ids) == amount_deleted, do: {:ok, amount_deleted}, else: {:error, "failed"}
```

Comment:
does this operation rollback the change in the db if it fails?

Reply:
Yup! You can check yourself by modifying the delete_units function to always return a wrong number:

```
def delete_units(unit_ids) do
  Repo.delete_all(from(u in Unit, where: u.id in ^unit_ids))
  {-1, nil}
end
```

Calling a "good" fusion (a simple one is 5-star Muflus as a target and 5 5-star Muflus as consumed units)

And checking the units count matches:
`user.id |> GameBackend.Users.get_user() |> Map.get(:units) |> Enum.count()` (run before and after the `Units.fuse()`)